File: Baekjoon/Class3/18.1012 유기농 배추.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split(/\s+/).map(Number);

const MOVE_DIR = [[-1, 0], [1, 0], [0, -1], [0, 1]]

let cursor = 1;
const countWorms = (out) => {
    const m = input[cursor++];
    const n = input[cursor++];
    const k = input[cursor++];
    
    const graph = Array.from({ length: m }, () => new Array(n).fill(false)); // 격자 그래프
    for(let _ = 0; _ < k; _++) {
        const r = input[cursor++];
        const c = input[cursor++];
        graph[r][c] = true;
    }

    let components = 0;
    for(let r = 0; r < m; r++) {
        for(let c = 0; c < n; c++) {
            if(graph[r][c]) { // 연결 요소 시작점 찾기
                graph[r][c] = false;
                components++;

                let bfsHead = 0;
                const bfs = [[r, c]];
                while(bfsHead < bfs.length) { // 너비 우선 탐색
                    const [startR, startC] = bfs[bfsHead++];
                    for(const [moveR, moveC] of MOVE_DIR) {
                        const endR = startR + moveR;
                        const endC = startC + moveC;
                        if(endR < 0 || endR >= m || endC < 0 || endC >= n)
                            continue;
                        if(graph[endR][endC]) {
                            bfs.push([endR, endC]);
                            graph[endR][endC] = false;
                        }
                    }
                }
            }
        }
    }
    out.push(components); // 연결 요소 수
}

const t = Number(input[0]);
const out = [];
for(let _ = 0; _ < t; _++)
    countWorms(out);
console.log(out.join("\n"));

File: Baekjoon/Class3/18.1012 유기농 배추.py
from collections import deque
import sys
input = sys.stdin.readline
write = sys.stdout.write

MOVE_DIR = ((-1, 0), (1, 0), (0, -1), (0, 1))

def count_worms():
    m, n, k = map(int, input().split())

    graph = [[False for _ in range(n)] for _ in range(m)] #격자 그래프
    for _ in range(k):
        r, c = map(int, input().split())
        graph[r][c] = True

    bfs = deque()
    components = 0
    for r in range(m):
        for c in range(n):
            if graph[r][c]: #연결 요소 시작점 찾기
                bfs.append((r, c))
                graph[r][c] = False
                components += 1
            while len(bfs): #너비 우선 탐색
                start_r, start_c = bfs.popleft()
                for move_r, move_c in MOVE_DIR:
                    end_r = start_r + move_r
                    end_c = start_c + move_c
                    if end_r < 0 or end_r >= m or end_c < 0 or end_c >= n:
                        continue
                    if graph[end_r][end_c]:
                        bfs.append((end_r, end_c))
                        graph[end_r][end_c] = False
    write(f"{components}\n") #연결 요소 수

def main():
    t = int(input())
    for _ in range(t):
        count_worms()

main()